src/metabase/feature_extraction/automagic_dashboards.clj

```diff
@@ -0,0 +1,350 @@
+                                                  :sizeY card-height}))
+  dashboard)
+
+(defn- build-query
+  [database context query]
+  (when-let [query (unify-vars context query)]
+    (let [database      (if (and (map? query)
+                                 (string? (:source_table query))
+                                 (s/starts-with? (:source_table query) "card__"))
+                          virtual-id
+                          database)
+          dataset_query (if (map? query)
+                          {:type     :query
+                           :query    query
+                           :database database}
+                          {:type     :native
+                           :native   {:query query}
+                           :database database})]
+      (when (perms/set-has-full-permissions-for-set?
+             @api/*current-user-permissions-set*
+             (card/query-perms-set dataset_query :write))
+        dataset_query))))
+
+(defn- create-card!
+  [database context {:keys [visualization title description query] :as card}]
+  (when-let [query (build-query database context query)]
+    (let [[visualization visualization-settings] (if (sequential? visualization)
+                                                   visualization
+                                                   [visualization {}])
+          card (db/insert! Card
+                 :creator_id             api/*current-user-id*
+                 :dataset_query          query
+                 :description            description
+                 :display                (or visualization :table)
+                 :name                   title
+                 :visualization_settings visualization-settings
+
+                 :result_metadata        (card.api/result-metadata-for-query query)
+                 :collection_id          nil)]
+      (events/publish-event! :card-create card)
+      (hydrate card :creator :dashboard_count :labels :can_write :collection)
+      card)))
+
+(defn- create-metrics!
+  [database context metrics]
+  (when api/*is-superuser?*
+    (into {}
+      (for [{:keys [description title metadata as query] :as metric} metrics]
+        [as (when-let [query (:query (build-query database context query))]
+              (metric/create-metric! (:source_table query)
+                                     title
+                                     (format "[Autogenerated] %s"
+                                             (or description ""))
+                                     api/*current-user-id*
+                                     query))]))))
+
+(def ^:private rules-dir "resources/automagic_dashboards")
+
+(defn- load-rules
+  []
+  (->> rules-dir
+       clojure.java.io/file
+       file-seq
+       (filter (memfn ^java.io.File isFile))
+       ; Workaround for https://github.com/owainlewis/yaml/issues/11
+       (map (comp yaml/parse-string slurp))))
+
+(defn populate-dashboards
```

Comment:
imo this should be two separate functions -- one to create dashboards for a table and one for a database. Passing in the "scope" here is too cute and short-sighted. It assumes the structure of of a database and table autogeneration function will be the same. 

Imo, it's probably a better pattern to have a single function that generates a dashboard for a table, and having a database generating function create a dashboard for every table of interest (and we should have a way to figure that out) in that database.